src/common/stats/StatsManager.cpp

```diff
@@ -309,6 +309,245 @@ StatusOr<StatsManager::VT> StatsManager::readHisto(const std::string& counterNam
+                                                 parts[1].c_str()));
+    }
+
+    return StatsManager::ParsedName {
+        name,
+        method,
+        range,
+    };
+}
+
+
+void PrometheusSerializer::annotate(
+        std::ostream& out, std::string& metric_name, std::string&& metric_type) const {
+    out << "# HELP " << metric_name << " Record all " << metric_type << " about nebula" << "\n";
+    out << "# TYPE " << metric_name << " " << metric_type << "\n";
+}
+
+// Write a double as a string, with proper formatting for infinity and NaN
+void PrometheusSerializer::writeValue(std::ostream& out, double value) const {
+    if (std::isnan(value)) {
+        out << "Nan";
+    } else if (std::isinf(value)) {
+        out << (value < 0 ? "-Inf" : "+Inf");
+    } else {
+        auto saved_flags = out.setf(std::ios::fixed, std::ios::floatfield);
+        out << value;
+        out.setf(saved_flags, std::ios::floatfield);
+    }
+}
+
+void PrometheusSerializer::writeValue(std::ostream& out, const int64_t value) const {
+    if (std::isnan(value)) {
+        out << "Nan";
+    } else if (std::isinf(value)) {
+        out << (value < 0 ? "-Inf" : "+Inf");
+    } else {
+        out << value;
+    }
+}
+
+void PrometheusSerializer::writeValue(std::ostream& out, const std::string& value) const {
+    for (auto c : value) {
+        switch (c) {
+        case '\n':
+            out << '\\' << 'n';
+            break;
+
+        case '\\':
+            out << '\\' << c;
+            break;
+
+        case '"':
+            out << '\\' << c;
+            break;
+
+        default:
+            out << c;
+            break;
+        }
+    }
+}
+
+template<typename T>
+void PrometheusSerializer::writeHead(std::ostream& out, const std::string& family,
+                const std::vector<Label>& labels, const std::string& suffix,
+                const std::string& extraLabelName,
+                const T extraLabelValue) const {
+    out << family << suffix;
+    if (!labels.empty() || !extraLabelName.empty()) {
+        out << "{";
+        const char* prefix = "";
+
+        for (auto& lp : labels) {
+            out << prefix << lp.name << "=\"";
+            writeValue(out, lp.value);
+            out << "\"";
+            prefix = ",";
+        }
+        if (!extraLabelName.empty()) {
+            out << prefix << extraLabelName << "=\"";
+            writeValue(out, extraLabelValue);
+            out << "\"";
+        }
+        out << "}";
+    }
+    out << " ";
+}
+
+// Write a line trailer: timestamp
+void PrometheusSerializer::writeTail(std::ostream& out, const std::int64_t timestamp_ms) const {
+    if (timestamp_ms != 0) {
+        out << " " << timestamp_ms;
+    }
+    out << "\n";
+}
+
+// TODO(shylock) transform nebula stats to Prometheus::ClientMetric/MetricFamily
+// using the serrializing from 3rd-SDK
+void StatsManager::prometheusSerialize(std::ostream& out) /*const*/ {
```

Comment:
Need some unit tests for this function

Reply:
Do you mean compare the output characters with the fixed characters we provide as answer?I don't think compare string is a good choice.

Reply:
Yes, add some.